tech/post/2017-11-10-Kubernetes-Affinity&AntiAffinity.md

```diff
@@ -0,0 +1,374 @@
+        topologyKey: failure-domain.beta.kubernetes.io/zone
+    podAntiAffinity:
+      preferredDuringSchedulingIgnoredDuringExecution:
+      - weight: 100
+        podAffinityTerm:
+          labelSelector:
+            matchExpressions:
+            - key: alias
+              operator: In
+              values:
+              - edison
+          topologyKey: kubernetes.io/hostname
+  containers:
+  - name: PodA
+    image: AppA
+```
+
+The Pod Affinity rule says that the Pod can be scheduled into a Node
+only if the Node has the same label ```failure-domain.beta.kubernetes.io/zone```,
+which has at least one already running Pod that has a Pod label with key-value ```alias``` in ```edison```.
+
+The Pod AntiAffinity rule says that the Pod will try not to be scheduled into a Node
+if the Node has the same label ```kubernetes.io/hostname```,
+which has at least one already running Pod that has a Pod label with key-value ```alias``` in ```edison```.
+
+Because ```PodA``` itself has a label with key-value ```alias``` = ```edison``.
+So this rule means that ```AppA``` is Affinity each other in the same Zone,
+and is AntiAffinity each other in the same Node.
+
+#### Example B
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: PodB
+  labels:
+    alias: xiang
+spec:
+  affinity:
+    podAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+      - labelSelector:
+          matchExpressions:
+          - key: alias
+            operator: In
+            values:
+            - edison
+        topologyKey: kubernetes.io/hostname
+    podAntiAffinity:
+      preferredDuringSchedulingIgnoredDuringExecution:
+      - weight: 100
+        podAffinityTerm:
+          labelSelector:
+            matchExpressions:
+            - key: alias
+              operator: In
+              values:
+              - xiang
+          topologyKey: kubernetes.io/hostname
+  containers:
+  - name: PodB
+    image: AppB
+```
+
+The Pod Affinity rule says that the Pod can be scheduled into a Node
+only if the Node has the same label ```kubernetes.io/hostname```,
+which has at least one already running Pod that has a Pod label with key-value ```alias``` in ```edison```.
+
+The Pod AntiAffinity rule says that the Pod will try not to be scheduled into a Node
+if the Node has the same label ```kubernetes.io/hostname```,
+which has at least one already running Pod that has a Pod label with key-value ```alias``` in ```xiang```.
+
+Because ```PodA``` has a label with key-value ```alias``` = ```edison``,
+and ```PodB``` itself has a label with key-value ```alias``` = ```xiang``,
+So this rule means that ```AppB``` is Affinity with ```AppA``` in the same Node,
+and is AntiAffinity itself in the same Node.
+
+### Pod Affinity and Pod AntiAffinity Note
+
+Pod Affinity and Pod AntiAffinity require very complex processing
+which can slow down scheduling in large clusters significantly.
+The Kubernetes do not recommend to use them more than 100 Nodes in clusters.

```

Comment:
Do you mean we should not use them if we have a cluster which has more than 100 Nodes?

Reply:
yeah, It will cost lots of resources to calculate.